modules/transformer-kinesis/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/kinesis/sinks/generic/Partitioned.scala

```diff
@@ -48,50 +36,82 @@ object Partitioned {
    * with `W`. [[SinkState]] can have (renewing over time) [[KeyedEnqueue]] that is when emitted
    *
    * @param getSink a sink constructor. For every new `W`, `K` and buffer we create a new sink
    *                which usually means a new file of key on blob store
    *                Can be taken from `sinks.s3` or `sinks.file`
-   * @param onComplete a callback executed after completion of a window
    */
-  def write[F[_]: Concurrent, W: Order, K: Show, V](getSink: W => K => Pipe[F, V, Unit], onComplete: W => F[Unit]): Pipe[F, Record[F, W, (K, V)], KeyedEnqueue[F, K, V]] =
-    (in: Partitioned[F, W, K, V]) =>
-      Stream.resource(Hotswap(SinkState.start[F, W, K, V](None, getSink))).flatMap {
-        case (initialHotswap, noopWindow) =>
-          def go(hotswap: WindowSwap[F, W, K, V],
-                 s: Partitioned[F, W, K, V],
-                 state: SinkState[F, W, K, V]): Pull[F, KeyedEnqueue[F, K, V], Unit] = {
-            s.pull.uncons1.flatMap {
-              // Initial, no-op window, exists only to get first _real_ window and use it for first enqueue
-              case Some((record, t)) if state.isNoop =>
-                rotate(getSink)(state, hotswap, true, Sync[F].unit, record.window)
-                  .flatMap(state => go(hotswap, Stream(record) ++ t, state))
-
-              // Still in the same window as with previous element - check if buffer can be emitted
-              case Some((Record.Data(_, checkpoint, (k, v)), t)) =>
-                for {
-                  _     <- Pull.eval(state.enqueue.enqueueKV(k, v))
-                  state <- emit(getSink)(state, checkpoint)
-                  _     <- go(hotswap, t, state.increment)
-                } yield ()
-
-              // New window
-              case Some((Record.EndWindow(completed, next, checkpoint), t)) =>
-                val finalize = onComplete(completed) >> checkpoint
-                for {
-                  newState  <- rotate(getSink)(state, hotswap, false, finalize, next)
-                  _         <- go(hotswap, t, newState)
-                } yield ()
-
-              case None =>
-                Pull.eval(logger.warn("Reached the end of the stream"))
-            }
+  def write[F[_]: Concurrent, W: Order: Show, K: Show, V, D: Monoid](getSink: W => D => K => Pipe[F, V, Unit]): Pipe[F, Record[W, List[(K, V)], D], (W, D)] = {
+    def go(s: Partitioned[F, W, K, V, D],
+           stateOpt: Option[SinkState[W, K, V, D]]): Pull[F, (W, D), Unit] = {
+      s.pull.uncons1.flatMap {
+
+        // Still in the same window as with previous element - check if buffer can be emitted
+        case Some((Record.Data(w, items, d), t)) =>
+          for {
+            state <- getState(stateOpt, w)
+            state <- items.foldM(state.addData(d)) { case (state, (k, v)) =>
+                       Pull.eval(state.enqueueKV(k, v))
+                     }
+            state <- maybeEmit[F, W, K, V, D](getSink, state)
```

Comment:
Yes, every line uses the one most recently created.

I quite like this pattern, because I find it less error-prone than the alternative:

```scala
for {
  state2 <- foo(state1)
  state3 <- bar(state2)
  state4 <- baz(state2) // Whoops!  wrong state
} yield state4
```

Another approach I suppose is to use a StateT monad.  I didn't think of that at the time.